src/object/obj_class.c

```diff
@@ -482,3 +496,76 @@ obj_ec_codec_get(daos_oclass_id_t oc_id)
 
 	return NULL;
 }
+
+
+/**
+ * Encode (using ISA-L) a full stripe from the submitted scatter-gather list.
+ *
+ * oid		[IN]		The object id of the object undergoing encode.
+ * sgl		[IN]		The SGL containing the user data.
+ * sg_idx	[IN|OUT]	Index of sg_iov entry in array.
+ * sg_off	[IN|OUT]	Offset into sg_iovs' io_buf.
+ * parity	[IN|OUT]	Struct containing parity buffers.
+ * p_idx	[IN]		Index into parity p_bufs array.
+ */
+int
+obj_encode_full_stripe(daos_obj_id_t oid, daos_sg_list_t *sgl, uint32_t *sg_idx,
+		       size_t *sg_off, struct obj_ec_parity *parity, int p_idx)
+{
+	struct obj_ec_codec		*codec = obj_ec_codec_get(
+							daos_obj_id2class(oid));
+	struct daos_oclass_attr		*oca = daos_oclass_attr_find(oid);
+	unsigned int			 len = oca->u.ec.e_len;
+	unsigned int			 k = oca->u.ec.e_k;
+	unsigned int			 p = oca->u.ec.e_p;
+	unsigned char			*data[k];
+	unsigned char			*ldata[k];
+	int				 i, lcnt = 0;
+	int				 rc = 0;
+
+	for (i = 0; i < k; i++)
+		if (sgl->sg_iovs[*sg_idx].iov_len - *sg_off >= len) {
+			unsigned char *from =
+				(unsigned char *)sgl->sg_iovs[*sg_idx].iov_buf;
+
+			data[i] = &from[*sg_off];
+			*sg_off += len;
+			if (*sg_off == sgl->sg_iovs[*sg_idx].iov_len) {
+				*sg_off = 0;
+				(*sg_idx)++;
+			}
+		} else {
+			int cp_cnt = 0;
+
+			D_ALLOC(ldata[lcnt], len);
+			if (ldata[lcnt] == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+			while (cp_cnt < len) {
+				int cp_amt =
+					sgl->sg_iovs[*sg_idx].iov_len-*sg_off <
+					len - cp_cnt ?
+					sgl->sg_iovs[*sg_idx].iov_len-*sg_off :
+					len - cp_cnt;
+				unsigned char *from =
+					sgl->sg_iovs[*sg_idx].iov_buf;
+
+				memcpy(&ldata[lcnt][cp_cnt], &from[*sg_off],
+				       cp_amt);
```

Comment:
I think need to pass full len to isa-l, the encode result is different if you do encode two times (for example pass first half and 2nd half). can refer the code of ec_encode_data_base(), will get different result as it get the gftable's data by (v[j * 32 + l * srcs * 32 + 1]). There are more advanced version based on sse/avx asm code, but should keep same result.

I considered to refer ec_encode_data_base()'s cod to write another version of ec_encode_data_sgl() to avoid the memcpy, but it is hard to change sse/avx version asm code (it is much much faster than the base version).
The memcpy here probably is fine for now.

Reply:
Oh, I do not mean encode multiple times for each byte, for example in some extreme case, you just moving forward pointer (data[i]) and parity pointer by 1 byte each time,  i.e. len = 1.  By this way, I think you do not need encoding the whole stripe by one ec_encode_data() call, and avoiding extra copy and allocation here.

Reply:
It looks like Xuezhao has checked and found that in general, the stripe cannot be vertically partitioned for RS.